### DISCUSSION TITLE
GPU Driver

### DISCUSSION THREAD

**Discussion by Liujiaqi-jlu** [gem5/gem5#1764] [Q&A] [answered]
Hello, I would like to ask about the drivers. Are the GPU drivers the same in APU mode and dGPU mode? Does gpu_compute_driver.hh correspond to the driver for each GPU? In src/dev/hsa/HSADriver.py, there is a mention of hsa_driver.hh, but there is no corresponding file in the folder. Will this affect HSA? Thank you!

**Comment by abmerop** [comment, marked as answer]
If you are using SE mode, the code in src/gpu-compute/gpu_compute_driver.cc is used. It does not matter if you are simulating APU or dGPU.

If you are using FS mode, the amdgpu driver in the Linux kernel is used.

HSADriver code was combined wtih gpu_compute_driver 5 years ago but it looks like the file still exists.

**Reply by Liujiaqi-jlu** [reply]
Thank you very much for your reply. I have another question regarding the AMDGPUDdevice. I followed the tutorial and used the apu_se.py configuration file to build the GPU in SE mode. I noticed that AMDGPUDdevice is a member variable of the GPUCommandProcessor class, but it is not instantiated within the GPUCommandProcessor class in apu_se.py. I also looked at the generated m5out/config.dot.pdf, and AMDGPUDdevice does not appear in the structure diagram. My question is: in SE mode, is the AMDGPUDdevice not instantiated as a physical device? I only found the instantiation process in the gem5/configs/example/gpufs/system/amdgpu.py file, but that should be in FS mode. I look forward to your guidance!


dispatcher = GPUDispatcher(kernel_exit_events=True)
gpu_cmd_proc = GPUCommandProcessor(hsapp=gpu_hsapp, dispatcher=dispatcher)
gpu_driver.device = gpu_cmd_proc
shader.dispatcher = dispatcher
shader.gpu_cmd_proc = gpu_cmd_proc

**Reply by mattsinc** [reply]
@Liujiaqi-jlu I'm pretty sure AMDGPUDevice is only instantiated in GPUFS, for the reasons @abmerop described above.

**Reply by abmerop** [reply]
apu_se.py will intercept the `ioctl` system call (what normally communicates with the driver) and will interface directly with the command processor in the GPU pipeline model.  Therefore it does not need all of the "uncore" GPU components.

In full system/using real driver, more things need to be modeled since the driver expects them (PM4, Interrupt Handler, GPU Virtual Memory, DMA engines).

**Reply by Sherry-yanqi** [reply]
Hi @abmerop ,
I am currently working on integrating my custom GPU instructions into the gem5 simulator, specifically for SE mode. My software stack involves OpenCL，Mesa and the DRM framework.
I would like to know if I need to add support for any PM4 (Packet-based Memory Management) related modules in SE mode to facilitate this integration. If need, how can I do this work? I'm looking forward your relpy.

**Reply by abmerop** [reply]
No, PM4 is only used in FS mode

**Reply by Sherry-yanqi** [reply]
But how can I finish my work? If I want to support DRM framework.
